### PR TITLE
Geant4 sensitive detectors cleanup 2

### PR DESCRIPTION
**Pull request by civanch**
Cleanup of main Sensitive detectors classes for calorimeters:
1) removed variables shadowing
2) reduce number of repeated computations
3) reduced computations when energy deposition zero
4) use of const pointers to G4 objects in private methods
5) after code checker applied extra files and modifications were added

Estimated speed-up of the SIM step ~1%, results should be identical.

### CODE REVIEW

**Comment by civanch** [comment]
please test

**Comment by civanch** [comment]
+1

File: SimG4CMS/Calo/interface/CaloSD.h [hunk 1]
```diff
@@ -49,23 +49,23 @@ class CaloSD : public SensitiveCaloDetector,
 
 public:    
   
-  CaloSD(G4String  aSDname, const DDCompactView & cpv,
+  CaloSD(std::string aSDname, const DDCompactView & cpv,
```
Comment (davidlange6) [inline SimG4CMS/Calo/interface/CaloSD.h]: hi @civanch - please make a const & string

File: SimG4Core/SensitiveDetector/src/SensitiveDetector.cc [hunk 2]
```diff
@@ -80,3 +81,10 @@ void SensitiveDetector::NaNTrap( G4Step* aStep )
   }
   return;
 }
+
+std::vector<std::string> SensitiveDetector::getNames()
+{
+  std::vector<std::string> temp;
```
Comment (davidlange6) [inline SimG4Core/SensitiveDetector/src/SensitiveDetector.cc]: return std::vector<std::string> temp(1, name);

**Comment by civanch** [comment]
please test

File: SimG4CMS/Calo/interface/CaloSD.h (hunk 1 above)
Comment (civanch) [inline SimG4CMS/Calo/interface/CaloSD.h]: Hi David, unfortunately I cannot because in the base G4 interface the string is not const....

**Comment by civanch** [comment]
+1

File: SimG4CMS/Calo/interface/CaloSD.h (hunk 1 above)
Comment (civanch) [inline SimG4CMS/Calo/interface/CaloSD.h]: Hi David, it is possible to use const std::string in CMSSW classes but the change requires addition of similar fix to 12 more files. It is done and I trigger tests now.

**Comment by civanch** [comment]
please test

**Comment by civanch** [comment]
+1

**Comment by davidlange6** [comment]
+1

**Comment by slava77** [comment]
PR description

> results should be identical

Comparisons:
> Reco comparison results: 1353 differences found in the comparisons

and I didn't find any notes in the thread justifying why this is OK.

File: SimG4Core/SensitiveDetector/interface/SensitiveDetector.h [hunk 3]
```diff
@@ -22,32 +21,29 @@ class DDCompactView;
...
   enum coordinates {WorldCoordinates, LocalCoordinates};
-  Local3DPoint InitialStepPosition(G4Step * step, coordinates);
-  Local3DPoint FinalStepPosition(G4Step * step, coordinates);
+  Local3DPoint InitialStepPosition(const G4Step * step, coordinates);
+  Local3DPoint FinalStepPosition(const G4Step * step, coordinates);
+
   inline Local3DPoint ConvertToLocal3DPoint(const G4ThreeVector& point)
   {
     Local3DPoint res(point.x(),point.y(),point.z());
     return std::move(res);
```
Comment (VinInn) [inline SimG4Core/SensitiveDetector/interface/SensitiveDetector.h]: do not understand where this idiom comes from: it is clearly stated a bit everywhere that one shall NOT move return values. it usually prevent RVO that is more efficient...

**Pull request merged**